### DISCUSSION TITLE
Type narrowing

### DISCUSSION THREAD

**Discussion by swastik** [typed-ember/glint#548] [Q&A] [answered]
Hi there! Another question. We have a component typed liked this (rough code):

```ts
interface Args {
  type: 'a' | 'b'
}
```

```hbs
{{#if (eq type 'a')}}
  <ComponentA @arg={{a}} />
{{/if}}
```

… and `ComponentA` is typed like this:

```ts
interface Args {
  type: 'a'
}
```

… I'll see an error saying `ComponentA` expects `a` and we're passing in `'a' | 'b'` — is there a way to fix this? Thank you!

**Comment by chriskrycho** [comment, marked as answer]
You need an implementation of `eq` which "narrows" the same way that TypeScript's `===` does. At the moment, the types for `eq` from `ember-truth-helpers` or similar do not *in general* support that. I think we're missing some docs on that, but you can see how it works to give it those semantics [in the implementing pull request](https://github.com/typed-ember/glint/pull/510). The reason this isn't "built in" is noted in the PR:

> Note that the type effects of these operators do not strictly match the semantics specified by RFC 562, in that they express **JavaScript truthiness** instead of **Handlebars truthiness**. It's not clear to me that it is possible to express Handlebars truthiness in TypeScript in such a way that you can usefully narrow types with it in combination with logical operators.

In other words, it's the same basic scenario as discussed in #547!
